File: tests/oracle/funcdata_mapglobals_maxvn_1204.cc
// FUNCDATA-MAPGLOBALS-MAXVN-0001 fixture — Funcdata::mapGlobals group
// maxvn carrying (R-MAPGLOBALS REJECT fix-forward discrimination).
//
// Locked oracle: Ghidra_12.0.4_build e40ed13014025f82488b1f8f7bca566894ac376b.
//
// Pinned oracle behavior (funcdata_varnode.cc:1653-1719): the inner group
// loop REASSIGNS maxvn on strictly greater size (cc:1685-1686
// `if (vn->getSize() > maxvn->getSize()) maxvn = vn;`), and the group's
// Datatype takes the biggest varnode's high type when it spans exactly
// [addr,endaddr) (cc:1692-1693 `ct = maxvn->getHigh()->getType();`).
// The same-address dual-width persist shape (loc order sorts size
// ascending — varnode.cc:41 — so the 1-byte varnode is the group start)
// is the input class the R-MAPGLOBALS review flagged as latent-UNTESTED:
// the group-start varnode's high type and the max varnode's high type are
// different observables.
//
// Cases (one stdout line each):
//   a_maxvn_type   1B@0x6000 (start, int1) + 8B@0x6000 (maxvn, forced
//                  4-byte uint high type): discovery arm — addSymbol's ct
//                  is the MAX varnode's high type (name base 'u', symbol
//                  type size 4, entry size 4), not the 8-byte span
//                  fallback and not the 1-byte start type.
//   b_entryflip    1B@0x6100 (start, forced 8-byte int high type) + 8B@0x6100
//                  (maxvn, forced 2-byte int high type) with a seeded 4-byte
//                  entry at 0x6100: ct size 2 keeps (addr+ct_size)-1 <=
//                  entry_end-1, so NO inconsistentuse / warningHeader (the
//                  group-start source would read the 8-byte start type and
//                  flip it — the review's entry-arm discrimination). The
//                  entry stays the seeded symbol.
//   c_fallback     8B@0x6200 (start) + 1B@0x6201 (internal, uncovered):
//                  endaddr shrinks to 0x6202 (cc:1684 unconditional),
//                  the maxvn span gate fails, ct = getBase(2,TYPE_UNKNOWN)
//                  — entry size 2, unknown metatype (name not printed:
//                  core-unknown NAMING differs between the fixture type
//                  factory and Rugra's shared factory; only structural
//                  fields are observed).
//   d_nomaxswap    8B@0x6300 (start, uint8) + 8B@0x6300 (equal size, int8):
//                  equal size never swaps maxvn (strictly-greater '>'):
//                  ct = the FIRST (start) varnode's type ('u', size 8).
//   e_multigroup   1B@0x6400 (start) + 8B@0x6400 (maxvn, uint4) + 4B@0x6404
//                  (internal uncovered, discovery arm ignores it): ct =
//                  maxvn's uint4 — same projection as (a) with the extra
//                  trailing member.
//   warn           count of warningheader comments after mapGlobals — 0 in
//                  the aligned state (the buggy span fallback would emit
//                  one from case b).
#include <bits/stdc++.h>

// Test-only access mirrors the house fixture pattern
// (tests/oracle/varnode_copy_symbol_1204.cc): production writes the persist
// bit through the localmap queryProperties channel; the fixture pins the
// varnode flag/type directly to construct the same bank state.
#define private public
#define protected public
#include "architecture.hh"
#include "comment.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef private
#undef protected

using namespace ghidra;

namespace {

class FixtureTranslate final : public Translate {
  VarnodeData dummy_register;

public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false, 8, 1, 1,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    AddrSpace *ram = new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false,
                                   8, 1, 3, AddrSpace::hasphysical, 0, 0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this, this, IPTR_PROCESSOR, "register", false,
                                   8, 1, 4, AddrSpace::hasphysical, 0, 0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(
        this, this, "stack", 5, 8, ram, 1, true);
    insertSpace(stack);
    VarnodeData stack_pointer;
    stack_pointer.space = reg;
    stack_pointer.offset = 0;
    stack_pointer.size = 8;
    addSpacebasePointer(stack, stack_pointer, 8, true);
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "join", false, 8, 1, 6,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new IopSpace(this, this, 7));
    setDefaultCodeSpace(3);
    dummy_register.space = reg;
    dummy_register.offset = 0;
    dummy_register.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    return dummy_register;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture() {
    FixtureTranslate *fixture_translate = new FixtureTranslate();
    translate = fixture_translate;
    copySpaces(fixture_translate);
    max_basetype_size = 16;
    types = new TypeFactory(this);
    types->setupSizes();
    // Core unknowns use the data-organization spelling (undefinedN) so the
    // mapGlobals span fallback getBase(n,TYPE_UNKNOWN) names identically to
    // Rugra's shared factory; the forced int/uint highs are the observable
    // discriminators (only the first character of a type name is observable
    // through buildVariableName's printNameBase).
    types->setCoreType("undefined1", 1, TYPE_UNKNOWN, false);
    types->setCoreType("undefined2", 2, TYPE_UNKNOWN, false);
    types->setCoreType("undefined4", 4, TYPE_UNKNOWN, false);
    types->setCoreType("undefined8", 8, TYPE_UNKNOWN, false);
    types->setCoreType("int1", 1, TYPE_INT, false);
    types->setCoreType("int2", 2, TYPE_INT, false);
    types->setCoreType("int8", 8, TYPE_INT, false);
    types->setCoreType("uint4", 4, TYPE_UINT, false);
    types->setCoreType("uint8", 8, TYPE_UINT, false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst, types, translate);
    symboltab = new Database(this, false);
    symboltab->attachScope(new ScopeInternal(0x101, "", this), (Scope *)0);
    commentdb = new CommentDatabaseInternal();
    ProtoModel *model = new ProtoModel(this);
    std::istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  void printMessage(const std::string &) const override {}
};

string metaname(type_metatype meta)
{
  switch (meta) {
  case TYPE_VOID: return "void";
  case TYPE_UNKNOWN: return "unknown";
  case TYPE_INT: return "int";
  case TYPE_UINT: return "uint";
  case TYPE_BOOL: return "bool";
  case TYPE_CODE: return "code";
  case TYPE_FLOAT: return "float";
  case TYPE_PTR: return "ptr";
  case TYPE_ARRAY: return "array";
  case TYPE_STRUCT: return "struct";
  case TYPE_UNION: return "union";
  default: return "other";
  }
}

} // namespace

int main(void)
{
  std::cout << std::unitbuf;
  // libdecomp.cc house pattern (cf. funcdata_assign_high_1204): registers
  // the PrintC capability the base Architecture ctor demands.
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);

  FixtureArchitecture architecture;
  AddrSpace *ram = architecture.translate->getSpaceByName("ram");
  Scope *global_scope = architecture.symboltab->getGlobalScope();

  Funcdata *fd = new Funcdata("mapglobals", "mapglobals",
                              global_scope, Address(ram, 0x5000),
                              (FunctionSymbol *)0, 0x100);
  BlockBasic *block = const_cast<BlockGraph &>(fd->getBasicBlocks()).newBlockBasic(fd);

  // Written persist varnode at (addr,size) with a forced Datatype.
  // cc:1671's group walk only needs written+persist+no-mapentry; the
  // forced type is what maxvn->getHigh()->getType() derives
  // (HighVariable::updateType, variable.cc:400-417, picks the single
  // member's type).
  int pc = 0x5010;
  auto make_persist_out = [&](int4 size, uintb addr,
                              type_metatype meta, int4 tsize) -> Varnode * {
    PcodeOp *op = fd->newOp(1, Address(ram, pc));
    pc += 1;
    fd->opSetOpcode(op, CPUI_COPY);
    Varnode *vn = fd->newVarnodeOut(size, Address(ram, addr), op);
    fd->opSetInput(op, fd->newConstant(size, 0x11), 0);
    fd->opInsertEnd(op, block);
    vn->type = architecture.types->getBase(tsize, meta);
    vn->setFlags(Varnode::persist | Varnode::addrtied);
    return vn;
  };

  Varnode *a1 = make_persist_out(1, 0x6000, TYPE_INT, 1);    // group start
  Varnode *a8 = make_persist_out(8, 0x6000, TYPE_UINT, 4);   // maxvn
  (void)a1; (void)a8;

  Varnode *b1 = make_persist_out(1, 0x6100, TYPE_INT, 8);    // start, 8-byte type
  Varnode *b8 = make_persist_out(8, 0x6100, TYPE_INT, 2);    // maxvn, 2-byte type
  (void)b1; (void)b8;

  Varnode *c8 = make_persist_out(8, 0x6200, TYPE_UINT, 8);   // group start
  Varnode *c1 = make_persist_out(1, 0x6201, TYPE_INT, 1);    // internal
  (void)c8; (void)c1;

  Varnode *d8a = make_persist_out(8, 0x6300, TYPE_UINT, 8);  // start, uint8
  Varnode *d8b = make_persist_out(8, 0x6300, TYPE_INT, 8);   // equal size, int8
  (void)d8a; (void)d8b;

  Varnode *e1 = make_persist_out(1, 0x6400, TYPE_INT, 1);    // group start
  Varnode *e8 = make_persist_out(8, 0x6400, TYPE_UINT, 4);   // maxvn
  Varnode *e4 = make_persist_out(4, 0x6404, TYPE_INT, 4);    // internal
  (void)e1; (void)e8; (void)e4;

  // Channel state AFTER varnode creation (case-b seed + global ownership
  // ranges for discoverScope, database.cc:1353-1366).
  architecture.symboltab->addRange(global_scope, ram, 0, 0xffffffffffffffffULL);
  global_scope->addSymbol("seed_b", architecture.types->getBase(4, TYPE_UINT),
                          Address(ram, 0x6100), Address());

  fd->setHighLevel();
  fd->mapGlobals();

  // Observation helper: the smallest containing entry through the same
  // queryProperties channel mapGlobals used (cc:1701).
  auto describe = [&](uintb addr) -> string {
    uint4 fl = 0;
    SymbolEntry *entry = fd->getScopeLocal()->queryProperties(
        Address(ram, addr), 1, Address(), fl);
    if (entry == (SymbolEntry *)0)
      return "none";
    ostringstream out;
    out << "name=" << entry->getSymbol()->getName()
        << "|size=" << dec << entry->getSize()
        << "|mt=" << metaname(entry->getSymbol()->getType()->getMetatype())
        << "|tsize=" << dec << entry->getSymbol()->getType()->getSize();
    return out.str();
  };

  std::cout << "case=a_maxvn_type|" << describe(0x6000) << "\n";
  std::cout << "case=b_entryflip|" << describe(0x6100) << "\n";
  std::cout << "case=c_fallback|" << describe(0x6200) << "\n";
  std::cout << "case=d_nomaxswap|" << describe(0x6300) << "\n";
  std::cout << "case=e_multigroup|" << describe(0x6400) << "\n";

  int warnings = 0;
  CommentSet::const_iterator iter, enditer;
  iter = architecture.commentdb->beginComment(fd->getAddress());
  enditer = architecture.commentdb->endComment(fd->getAddress());
  for (; iter != enditer; ++iter) {
    if (((*iter)->getType() & Comment::warningheader) != 0)
      warnings += 1;
  }
  std::cout << "case=warn|count=" << dec << warnings << "\n";
  return 0;
}

File: tests/oracle/funcdata_mapglobals_maxvn_1204.rs
//! FUNCDATA-MAPGLOBALS-MAXVN-0001 Rugra comparand: `Funcdata::map_globals`
//! group maxvn carrying (funcdata_varnode.cc:1653-1719) against the locked
//! Ghidra 12.0.4 oracle. Mirrors tests/oracle/funcdata_mapglobals_maxvn_1204.cc
//! case by case; every printed value must byte-match the oracle.
//!
//! The pinned semantics are the inner group loop's maxvn reassignment
//! (cc:1685-1686, strictly-greater size carries the varnode Arc) and the
//! ct read (cc:1692-1693, the biggest varnode's high type when it spans
//! exactly [addr,endaddr)) — the R-MAPGLOBALS REJECT fix-forward
//! discrimination on the same-address dual-width persist shape.
//!
//! Structural mapping notes (registered in the metadata):
//!   * The fixture pins `persist|addrtied` and the forced `v_type` directly
//!     on written varnodes (production derives them from the localmap
//!     queryProperties channel in `Funcdata::newVarnodeOut`, a registered
//!     Rugra gap); the C++ twin does the same through test-only access.
//!   * The global-scope query channel is the Database (Rugra models the
//!     global scope as the query point; the C++ twin queries through the
//!     ScopeLocal whose parent walk reaches the same entries).
//!   * Core unknowns are named `undefinedN` on this side (data-organization
//!     flavor) — the C++ twin registers the same spelling so the span
//!     fallback's printNameBase contribution matches.
//!   * The 4-byte unsigned base is named `uint` here and `uint4` in the
//!     C++ twin; only the first character (`u`) is observable through
//!     buildVariableName's printNameBase, so created-symbol names match.
//!   * Case b forces an 8-byte type on the 1-byte group start: the
//!     maxvn-source fix reads the MAX member's 2-byte type (no overflow,
//!     no warning), while the pre-fix group-start source would read the
//!     8-byte start type and flip inconsistentuse/warningHeader.
use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::comment::comment_type;
use rugra::database::Database;
use rugra::funcdata::Funcdata;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::TypeMetatype;
use rugra::type_system::typefactory::TypeFactory;
use rugra::varmap::ScopeLocal;
use rugra::varnode::varnode_flags;

fn forced_type(size: usize, meta: TypeMetatype) -> Arc<rugra::type_system::datatype::Datatype> {
    TypeFactory::shared_default()
        .read()
        .unwrap()
        .get_base(size, meta)
        .expect("shared factory produces every base type")
}

fn metaname(meta: TypeMetatype) -> &'static str {
    use TypeMetatype::*;
    match meta {
        Unknown => "unknown",
        Void => "void",
        Bool => "bool",
        Int => "int",
        Uint => "uint",
        Float => "float",
        Pointer => "ptr",
        Array => "array",
        Struct => "struct",
        Union => "union",
        Enum => "other",
        Code => "code",
        Spacebase => "other",
        PartialStruct | PartialEnum | PartialUnion => "other",
    }
}

fn main() {
    // The channel: Database with a global scope (fresh — no symbols yet,
    // ranges seeded after varnode creation exactly like the C++ twin) and
    // a comment db so warningHeader is observable instead of stderr noise.
    let mut arch = rugra::arch::Architecture::new();
    let db_arc = Arc::new(RwLock::new(Database::new(false)));
    arch.set_symboltab(db_arc.clone());
    arch.commentdb = Some(Arc::new(RwLock::new(
        rugra::comment::CommentDatabaseInternal::new(),
    )));

    let mut fd = Funcdata::new("mapglobals", Address::new(0x5000), 0x100);
    fd.set_arch(Arc::new(arch));
    fd.scope = Some(ScopeLocal::new());
    let block = fd.create_new_block();

    // Written persist varnode at (addr,size) with a forced Datatype — the
    // same construction as the C++ twin's make_persist_out. The varnode is
    // created directly in the ram space (Rugra's new_varnode_out maps a
    // spaceless Address into the register space, which would take the
    // legacy proxy arm instead of the RAM query channel mapGlobals uses);
    // the HighVariable attaches during the set_high_level sweep below,
    // exactly like the C++ twin (highlevel_on is clear at creation).
    let mut pc = 0x5010u64;
    let mut make_persist_out = |fd: &mut Funcdata,
                                block: &Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
                                size: usize,
                                addr: u64,
                                tsize: usize,
                                meta: TypeMetatype| {
        let op = fd.new_op(1, Address::new(pc));
        pc += 1;
        fd.op_set_opcode(&op, OpCode::CPUI_COPY);
        let vn = fd.vbank.create_def_with_space(size, AddressSpace::Ram, addr, &op.0);
        op.0.write().unwrap().output = Some(vn.clone());
        let c = fd.new_constant(size, 0x11);
        fd.op_set_input(&op, c, 0);
        fd.op_insert_end(&op, block);
        let ct = forced_type(tsize, meta);
        let mut vn_w = vn.write().unwrap();
        vn_w.v_type = Some(ct);
        vn_w.set_flags(varnode_flags::PERSIST | varnode_flags::ADDRTIED);
        drop(vn_w);
        vn
    };

    let _a1 = make_persist_out(&mut fd, &block, 1, 0x6000, 1, TypeMetatype::Int);
    let _a8 = make_persist_out(&mut fd, &block, 8, 0x6000, 4, TypeMetatype::Uint);

    let _b1 = make_persist_out(&mut fd, &block, 1, 0x6100, 8, TypeMetatype::Int);
    let _b8 = make_persist_out(&mut fd, &block, 8, 0x6100, 2, TypeMetatype::Int);

    let _c8 = make_persist_out(&mut fd, &block, 8, 0x6200, 8, TypeMetatype::Uint);
    let _c1 = make_persist_out(&mut fd, &block, 1, 0x6201, 1, TypeMetatype::Int);

    let _d8a = make_persist_out(&mut fd, &block, 8, 0x6300, 8, TypeMetatype::Uint);
    let _d8b = make_persist_out(&mut fd, &block, 8, 0x6300, 8, TypeMetatype::Int);

    let _e1 = make_persist_out(&mut fd, &block, 1, 0x6400, 1, TypeMetatype::Int);
    let _e8 = make_persist_out(&mut fd, &block, 8, 0x6400, 4, TypeMetatype::Uint);
    let _e4 = make_persist_out(&mut fd, &block, 4, 0x6404, 4, TypeMetatype::Int);

    // Channel state AFTER varnode creation: global ownership range for
    // discover_scope (database.cc:1353-1366) + the case-b seeded entry.
    {
        let mut db = db_arc.write().unwrap();
        let global = db.global_scope_id;
        if let Some(rng) = rugra::address::Range::new(Address::new(0), Address::new(u64::MAX)) {
            db.add_range(global, rng);
        }
        let seed_type = forced_type(4, TypeMetatype::Uint);
        db.add_symbol_mapped(global, "seed_b", Some(seed_type), Address::new(0x6100), 4);
    }

    fd.set_high_level();
    fd.map_globals().expect("map_globals succeeds");

    // Observation helper: the same query channel map_globals used
    // (cc:1701 queryProperties(addr, 1, empty usepoint)).
    let describe = |fd: &Funcdata, addr: u64| -> String {
        match fd.query_properties_parent_scope(Address::new(addr), 1, Address::new(0)) {
            Some((Some(hit), _)) => format!(
                "name={}|size={}|mt={}|tsize={}",
                hit.symbol_name,
                hit.entry_size,
                metaname(hit.type_metatype),
                hit.symbol_type
                    .as_ref()
                    .map(|t| t.get_size())
                    .unwrap_or(0),
            ),
            _ => "none".to_string(),
        }
    };

    println!("case=a_maxvn_type|{}", describe(&fd, 0x6000));
    println!("case=b_entryflip|{}", describe(&fd, 0x6100));
    println!("case=c_fallback|{}", describe(&fd, 0x6200));
    println!("case=d_nomaxswap|{}", describe(&fd, 0x6300));
    println!("case=e_multigroup|{}", describe(&fd, 0x6400));

    let warnings = fd
        .arch
        .as_ref()
        .and_then(|a| a.commentdb.clone())
        .map(|cdb| {
            cdb.read()
                .unwrap()
                .comments_for_function(fd.baseaddr)
                .filter(|c| (c.get_type() & comment_type::WARNINGHEADER) != 0)
                .count()
        })
        .unwrap_or(0);
    println!("case=warn|count={warnings}");
}
